saleor/graphql/product/types/products.py

```diff
@@ -1090,37 +1148,54 @@ def resolve_image_by_id(root: ChannelContext[models.Product], _info, *, id):
 
     @staticmethod
-    def resolve_media(root: ChannelContext[models.Product], info):
-        return MediaByProductIdLoader(info.context).load(root.node.id)
+    def resolve_media(root: ChannelContext[models.Product], info, sort_by=None):
+        if sort_by is None:
+            sort_by = {
+                "field": MediaChoicesSortField.ID.value,  # type: ignore
+                "direction": "",
+            }
+        return MediaByProductIdLoaderSort(info.context, sort_by).load(root.node.id)
 
     @staticmethod
     def resolve_images(root: ChannelContext[models.Product], info):
         return ImagesByProductIdLoader(info.context).load(root.node.id)
 
     @staticmethod
-    def resolve_variants(root: ChannelContext[models.Product], info):
-        requestor = get_user_or_app_from_context(info.context)
-        has_required_permissions = has_one_of_permissions(
-            requestor, ALL_PRODUCTS_PERMISSIONS
-        )
-        if has_required_permissions and not root.channel_slug:
-            variants = ProductVariantsByProductIdLoader(info.context).load(root.node.id)
-        elif has_required_permissions and root.channel_slug:
-            variants = ProductVariantsByProductIdAndChannel(info.context).load(
-                (root.node.id, root.channel_slug)
-            )
-        else:
-            variants = AvailableProductVariantsByProductIdAndChannel(info.context).load(
-                (root.node.id, root.channel_slug)
+    def resolve_variant(root: ChannelContext[models.Product], info, id=None, sku=None):
+        if id is None and sku is None:
+            return None
+
+        keys = {}
+        if id is not None:
+            _, pk = from_global_id_or_error(id, ProductVariant)
+            keys["id"] = int(pk)
+
+        if sku is not None:
+            keys["sku"] = sku
+
+        def get_product_variant_by_id_and_sku(
+            product_variants, keys
+        ) -> Optional[ProductVariant]:
+            def condition(variant):
+                con = True
+                for key, val in keys.items():
+                    if val is None:
+                        con &= getattr(variant, key) is None
+                    else:
+                        con &= getattr(variant, key) == val
+                return con
+
+            return next(
+                (variant for variant in product_variants if condition(variant.node)),
+                None,
             )
 
-        def map_channel_context(variants):
-            return [
-                ChannelContext(node=variant, channel_slug=root.channel_slug)
-                for variant in variants
-            ]
+        variants = resolve_variants(root, info)
+        return variants.then(lambda x: get_product_variant_by_id_and_sku(x, keys))
 
-        return variants.then(map_channel_context)
+    @staticmethod
+    def resolve_variants(root: ChannelContext[models.Product], info):
+        return resolve_variants(root, info)
```

Comment:
Not sure about extracting resolver outside the type. What do we get by extracting it outside the type? I see that you use it also in `resolve_variant` but you could also call `Product.resolve_variants(root, info)`, right?